ckanext/canada/triggers.py

```diff
@@ -110,12 +125,39 @@ def update_triggers():
             BEGIN
                 IF value IS NOT NULL AND value <> '' AND NOT (value = ANY (choices)) THEN
                     -- \t is used when converting errors to string
-                    RETURN ARRAY[[field_name, 'Invalid choice: "'
+                    RETURN ARRAY[[field_name, 'Invalid choice: {}\uF8FF"'
                         || replace(value, E'\t', ' ') || '"']];
                 END IF;
                 RETURN NULL;
             END;
         ''')
+    # return record with .clean (trimmed value) and .error
+    # (NULL or ARRAY[[value_trimmed]])
+    # .clean must be unnest(.clean) from the return to get the text value
+    lc.action.datastore_function_create(
+        name=u'max_char_error',
+        or_replace=True,
+        arguments=[
+            {u'argname': u'value', u'argtype': u'text'},
+            {u'argname': u'max_chars', u'argtype': u'numeric'},
+            {u'argname': u'field_name', u'argtype': u'text'},
+            {u'argname': u'clean', u'argtype': u'_text', u'argmode': u'out'},
+            {u'argname': u'error', u'argtype': u'_text', u'argmode': u'out'}],
+        rettype=u'record',
+        definition=u'''
+            DECLARE
+                value_trimmed text := trim(both E'\t\n\x0b\x0c\r ' from value);
+            BEGIN
+                IF value IS NOT NULL AND value <> '' AND LENGTH(value_trimmed) > max_chars THEN
+                    error := ARRAY[[field_name, 'This field has a maximum length of {} characters.\uF8FF' || max_chars]];
+                END IF;
+                IF value IS NULL OR value = '' THEN
+                    clean := NULL;
+                ELSE
+                    clean := ARRAY(SELECT value_trimmed);
```

Comment:
I don't understand why this is needed

Reply:
the null part? or the trigger in general haha?

Will `value_trimmed` be null/empty string and we can just do the `clean := ARRAY(SELECT value_trimmed)`

Reply:
I mean the `ARRAY(SELECT value_trimmed)`, why can't we use `value_trimmed` directly?

Reply:
Also it might be simpler to make the trimming and error generation separate functions that are called from the trigger one after the next.

Reply:
move trimmed to its own function. update yaml now.